language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java

```diff
@@ -0,0 +1,443 @@
+      } catch (IOException e) {
+        parseExpected("valid textual value");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  // Read an list with an unknown number of items of the same type.
+  public <T> FromJson<List<T>> list(FromJson<T> readItem) {
+    return () -> {
+      List<T> list = new ArrayList<>();
+      readStartArray();
+      while (notEndArray()) {
+        T item = readItem.read();
+        list.add(item);
+      }
+      readEndArray();
+      return list;
+    };
+  }
+
+  // Read a map with textual keys, and unknown number of items of the same type.
+  public <V> FromJson<Map<String, V>> textMap(FromJson<V> readValue) {
+    return () -> {
+      Map<String, V> map = new TreeMap<>();
+      readStartObject();
+      while (notEndObject()) {
+        String key = readFieldName();
+        V val = readValue.read();
+        map.put(key, val);
+      }
+      ;
+      readEndObject();
+      return map;
+    };
+  }
+
+  // Read a map with unknown number of items of the same types.
+  public <K, V> FromJson<Map<K, V>> genMap(FromJson<K> readKey, FromJson<V> readValue) {
+    return () -> {
+      Map<K, V> map = new TreeMap<>();
+      // Maps are represented as an array of 2-element arrays.
+      readStartArray();
+      while (notEndArray()) {
+        readStartArray();
+        K key = readKey.read();
+        V val = readValue.read();
+        readEndArray();
+        map.put(key, val);
+      }
+      ;
+      readEndArray();
+      return map;
+    };
+  }
+
+  // The T type should not itself be Optional<?>. In that case use OptionalNested below.
+  public <T> FromJson<Optional<T>> optional(FromJson<T> readValue) {
+    return () -> {
+      if (parser.currentToken() == JsonToken.VALUE_NULL) {
+        moveNext();
+        return Optional.empty();
+      } else {
+        return Optional.of(readValue.read());
+      }
+    };
+  }
+
+  public <T> FromJson<Optional<Optional<T>>> optionalNested(FromJson<Optional<T>> readValue) {
+    return () -> {
+      if (parser.currentToken() == JsonToken.VALUE_NULL) {
+        moveNext();
+        return Optional.empty();
+      } else {
+        readStartArray();
+        Optional<T> val = notEndArray() ? readValue.read() : Optional.empty();
+        readEndArray();
+        return Optional.of(val);
+      }
+    };
+  }
+
+  public <E extends Enum<E>> FromJson<E> enumeration(Class<E> enumClass) {
+    return () -> {
+      String value = text().read();
+      try {
+        return Enum.valueOf(enumClass, value);
+      } catch (IllegalArgumentException e) {
+        parseExpected(String.format("constant of %s", enumClass.getName()));
+      }
+      return null;
+    };
+  }
+
+  // Provides a generic way to read a variant type, by specifying each tag.
+  public <T> FromJson<T> variant(List<String> tagNames, TagReader<T> readTag) {
+    return () -> {
+      readStartObject();
+      if (!readFieldName().equals("tag")) parseExpected("tag field");
```

Comment:
I might be misunderstanding how jackson works here but to me this looks like you're making assumptions on the order of keys. Specifically you seem to assume that "tag" always comes before "value". Your tests also only cover this case. That doesn't seem right to me. JSON objects are generally unordered an a lot of encoders don't even give you control over the field order.

Reply:
I'd suggest we fix it before merging.

Reply:
Done